File: python/med/q20.py
"""
You are tasked to debug a friend recommendation system to calculate mutual friends based on a unique 
user ID in an UNDIRECTED GRAPH.

Core Concepts:
- `defaultdict(set)`: Represents the adjacency list. Each key is a user ID, and the value is a `set` 
  of friends.
- `set`: Efficiently stores unique friends and allows quick lookups.
- `list[tuple[int, int]]`: Stores recommendations as tuples `(friendID, mutualFriendCount)`. The list 
  is sorted in descending order of mutual friends.

Key Python Methods:
- `.add`: Adds a unique element to a `set`.
- `in`: Checks if an element exists in a `set`.
- `sorted`: Sorts the recommendation list using `key=lambda x: x[1]` for mutual friends, 
  with `reverse=True` for descending order.

Expected Output:
- A correct listing of vertex : vertices connected
- A correct sorted list of friend recommendations based on the number of mutual friends.
"""

class FriendGraph:
    def __init__(self):
        self.graph = {}

    def add_user(self, user):
        self.graph[user] = set()

    def add_friendship(self, user1, user2):
        if user1 in self.graph and user2 in self.graph:
            self.graph[user1].add(user2)

    def recommend_friends(self, user):
        recommendation_count = {}
        print(f"Calculating friend recommendations for User {user}")

        for other_user, friends in self.graph.items():

            mutual_friends = sum(1 for friend_of_user in self.graph[user] if friend_of_user in friends)

            recommendation_count[other_user] = mutual_friends
            print(f"User {other_user} has {mutual_friends} mutual friends with User {user}")

        recommendations = sorted(recommendation_count.items(), key=lambda x: -x[1])
        return recommendations

    def display_graph(self):
        for user, friends in self.graph.items():
            if friends:
                print(f"User {user}: {' '.join(map(str, friends))}")


graph = FriendGraph()
graph.add_user(1)
graph.add_user(2)
graph.add_user(3)
graph.add_user(4)
graph.add_user(5)
graph.add_user(6)
graph.add_friendship(1, 2)
graph.add_friendship(6, 2)
graph.add_friendship(1, 3)
graph.add_friendship(2, 4)
graph.add_friendship(3, 4)

print("Graph:")
graph.display_graph()

print("\nFriend Recommendations for User 1:")
recommendations = graph.recommend_friends(1)
for user, count in recommendations:
    print(f"User {user} (Mutual Friends: {count})")

File: js/med/q20.js
/**
 * You are tasked to debug a friend recommendation system to calculate mutual friends based on a unique 
 * user ID in an UNDIRECTED GRAPH.
 * 
 * Core Concepts:
 * - `Map`: Represents the adjacency list. Each key is a user ID, and the value is a `Set` of friends.
 * - `Set`: Efficiently stores unique friends and allows quick lookups.
 * - `Array`: Stores recommendations as `[friendID, mutualFriendCount]`. The array is sorted in descending 
 *   order of mutual friends.

 * Key JavaScript Methods:
 * - `.add`: Adds a unique element to a `Set`.
 * - `.has`: Checks if an element exists in a `Set`.
 * - `.sort`: Sorts the recommendation list. Use a custom comparison function for descending order.

 * Expected Output:
 * - A correct listing of vertex : vertices connected
 * - A correct sorted array of friend recommendations based on the number of mutual friends.
 */


class FriendGraph {
    constructor() {
        this.graph = new Map();
    }

    addUser(user) {
        this.graph.set(user, new Set());
    }

    addFriendship(user1, user2) {
        if (this.graph.has(user1) && this.graph.has(user2)) {
            this.graph.get(user1).add(user2);
        }
    }

    recommendFriends(user) {
        const recommendationCount = new Map();
        console.log(`Calculating friend recommendations for User ${user}`);

        for (const [otherUser, friends] of this.graph.entries()) {

            let mutualFriends = 0;
            for (const friendOfUser of this.graph.get(user)) {
                if (friends.has(friendOfUser)) {
                    mutualFriends++;
                }
            }

            recommendationCount.set(otherUser, mutualFriends);
            console.log(`User ${otherUser} has ${mutualFriends} mutual friends with User ${user}`);
        }

        const recommendations = Array.from(recommendationCount.entries())
            .sort((a, b) => b[1] - a[1]);
        return recommendations;
    }

    displayGraph() {
        for (const [user, friends] of this.graph.entries()) {
            if (friends.size > 0) {
                console.log(`User ${user}: ${[...friends].join(' ')}`);
            }
        }
    }
}

const graph = new FriendGraph();
graph.addUser(1);
graph.addUser(2);
graph.addUser(3);
graph.addUser(4);
graph.addUser(5);
graph.addUser(6);
graph.addFriendship(1, 2);
graph.addFriendship(6, 2);
graph.addFriendship(1, 3);
graph.addFriendship(2, 4);
graph.addFriendship(3, 4);

console.log("Graph:");
graph.displayGraph();

    // Expected Output (flexible):
    /*
     *  User 6: 2
        User 5:
        User 4: 3 2 
        User 3: 4 1 
        User 2: 4 1 
        User 1: 3 2
    */
    

console.log("\nFriend Recommendations for User 1:");
const recommendations = graph.recommendFriends(1);
for (const [user, count] of recommendations) {
    console.log(`User ${user} (Mutual Friends: ${count})`);
}


// Expected Output:
    /**
     * Friend Recommendations for User 1:
     * Calculating Friend Recommendations for User 1
     * User 4 has 2 mutual friends with User 1
     * User 6 has 1 mutual friends with User 1
     * User 4 (Mutual Friends: 2)
     * User 6 (Mutual Friends: 1)
     */